src/tools/databricks.py

```diff
@@ -0,0 +1,298 @@
+                # Personal Access Token (PAT) authentication
+                logger.info("Using Personal Access Token (PAT) authentication")
+                self.session = sql.connect(
+                    server_hostname=host,
+                    http_path=http_path,
+                    access_token=access_token,
+                    catalog=catalog if catalog else None,
+                    schema=schema if schema else None,
+                    _enable_connection_pooling=True,  # Enable connection pooling
+                )
+
+            elif client_id and client_id.strip() and client_secret and client_secret.strip():
+                # M2M OAuth authentication
+                logger.info("Using M2M OAuth authentication")
+                self.session = sql.connect(
+                    server_hostname=host,
+                    http_path=http_path,
+                    auth_type="databricks-oauth",
+                    client_id=client_id,
+                    client_secret=client_secret,
+                    catalog=catalog if catalog else None,
+                    schema=schema if schema else None,
+                    _enable_connection_pooling=True,  # Enable connection pooling
+                )
+
+            else:
+                raise Exception(
+                    "No valid authentication method found. Provide either access_token or both client_id and client_secret"
+                )
+
+        except Exception as e:
+            raise Exception(f"Failed to create Databricks connection: {str(e)}")
+
+        return self.session
+
+    def ensure_valid_session(self) -> None:
+        """Ensure we have a valid Databricks connection."""
+        if self.session is None:
+            raise Exception(
+                "Connection is not initialized. Call initialize_warehouse_connection() mcp tool first."
+            )
+
+        try:
+            # Test the connection with a simple query
+            cursor = self.session.cursor()
+            cursor.execute("SELECT 1 as test_column")
+            cursor.fetchall()
+            cursor.close()
+            self.update_last_used()
+
+        except Exception as e:
+            # Connection is invalid, create new one
+            logger.warning(f"Databricks connection invalid or expired: {str(e)}")
+            if self.session is not None:
+                try:
+                    self.session.close()
+                except:
+                    pass
+
+            logger.info("Creating new Databricks connection due to expiration/invalidity")
+            self.session = self.create_session()
+            self.update_last_used()
+
+    def raw_query(
+        self, query: str, response_type: str = "list"
+    ) -> Union[List[Dict], pd.DataFrame]:
+        """Execute Databricks SQL query and return results."""
+        try:
+            logger.info(f"Executing Databricks query: {query[:100]}...")
+            self.ensure_valid_session()
```

Comment:
Need this for every query?

Reply:
This is good to be in mcp output, it can help us in identifying issues.